uvtools/dspec.py

```diff
@@ -146,6 +147,250 @@ def delay_filter(data, wgts, bl_len, sdf, standoff=0., horizon=1., min_dly=0.0,
+        the highest harmonic mode in the model.
+    
+    Nfreqs : int
+        Number of frequency channels to model. The Fourier modes are integer 
+        harmonics within this frequency window.
+    
+    Returns
+    -------
+    model : array_like
+        Fourier model constructed from the input harmonic coefficients. 
+        Shape: (Nfreqs,).
+    """
+    # Frequency array (*not* in physical frequency units)
+    nu = np.arange(Nfreqs)
+    L = nu[-1] - nu[0]
+    
+    assert(len(cn.shape) == 1), "cn must be a 1D array"
+    nmax = (cn.size - 1) // 2 # Max. harmonic
+    
+    # Build matrix operator for complex Fourier basis
+    n = np.arange(-nmax, nmax+1)
+    F = np.array([np.exp(-1.j * _n * nu / L) for _n in n])
+    
+    # Return model
+    return np.dot(cn, F)
+
+
+def delay_filter_leastsq_1d(data, flags, sigma, nmax, add_noise=False, 
+                            cn_guess=None, use_linear=True, operator=None):
+    """
+    Fit a smooth model to 1D complex-valued data with flags, using a linear 
+    least-squares solver. The model is a Fourier series up to a specified 
+    order. As well as calculating a best-fit model, this will also return a 
+    copy of the data with flagged regions filled in ('in-painted') with the 
+    smooth solution. 
+    
+    Optionally, you can also add an uncorrelated noise realization on top of 
+    the smooth model in the flagged region.
+    
+    Parameters
+    ----------
+    data : array_like, complex
+        Complex visibility array as a function of frequency, with shape 
+        (Nfreqs,).
+    
+    flags : array_like, bool
+        Boolean flags with the same shape as data.
+    
+    sigma : float or array_like
+        Noise standard deviation, in the same units as the data. If float, 
+        assumed to be homogeneous in frequency. If array_like, must have 
+        the same shape as the data.
+        
+        Note that the choice of sigma will have some bearing on how sensitive 
+        the fits are to small-scale variations.
+    
+    nmax: int
+        Max. order of Fourier modes to fit. A model with complex Fourier modes 
+        between [-n, n] will be fitted to the data, where the Fourier basis 
+        functions are ~ exp(-i 2 pi n nu / (Delta nu).
+    
+    add_noise : bool, optional
+        Whether to add an unconstrained noise realization to the in-painted areas. 
+        This uses sigma to set the noise standard deviation. Default: False.
+    
+    cn_guess : array_like, optional
+        Initial guess for the series coefficients. If None, zeros will be used. 
+        A sensible choice of cn_guess can speed up the solver significantly. 
+        Default: None.
+    
+    use_linear : bool, optional
+        Whether to use a fast linear least-squares solver to fit the Fourier 
+        coefficients, or a slower generalized least-squares solver. 
+        Default: True.
+    
+    operator : array_like, optional
```

Comment:
Because this argument is optional, you should mention what the default value is. ~~Also, do we want to have the option of letting the user specify the basis? Put another way, are there well-motivated reasons to _not_ use the regular Fourier basis? It seems like it could get confusing to keep track of when trying to reconstruct things later, especially since the produced coefficients depend on the basis passed in.~~ Disregard this, see comment below on the 2d version.